picas/actors.py

```diff
@@ -165,62 +171,57 @@ def cleanup_env(self, *args, **kwargs):
         Method which gets called after the run method has completed.
         """
 
-        
+
 class RunActor(AbstractRunActor):
     """
     RunActor class with added stopping functionality.
     """
 
-    def run(self, max_token_time=None, max_time=None, avg_time_factor=0.0, max_tasks=0, stop_function=None, **stop_function_args):
+    def run(self, max_token_time=None, max_total_time=None, max_tasks=None, max_scrub=0,
+            stop_function=None, **stop_function_args):
         """
         Run method of the actor, executes the application code by iterating
         over the available tasks in CouchDB, including stop logic. The stop
         logic is also extended into the EndlessViewIterator to break it when
         the condition is met, otherwise it never stops.
 
         @param max_token_time: maximum time to run a single token before stopping
-        @param max_time: maximum time to run picas before stopping
-        @param avg_time_factor: used for estimating when to stop with `max_time`,
-                                value is average time per token to run
+        @param max_total_time: maximum time to run picas before stopping
         @param max_tasks: number of tasks that are performed before stopping
+        @param max_scrub: number of times a token can be reset ('scrubbed') after failing
         @param stop_function: custom function to stop the execution, must return bool
         @param stop_function_args: kwargs to supply to stop_function
         """
-        self.time = Timer()
+        timer = Timer()
         self.prepare_env()
 
         # handler needs to be setup in overwritten method
         self.setup_handler()
 
-        # Special case to break the while loop of the EndlessViewIterator:
-        # The while loop cant reach the stop condition in the for loop below,
-        # so pass the condition into the stop mechanism of the EVI, then the
-        # iterator is stopped from EVI and not the RunActorWithStop
-        if isinstance(self.iterator, EndlessViewIterator):
-            self.iterator.stop_callback = stop_function
-            self.iterator.stop_callback_args = stop_function_args
+        # Break the while loop of the Iterator if max_total_time is exceeded
+        if max_total_time is not None:
```

Comment:
This only works if the Iterator is EndlessView, the other Iterators do not have the `stop_function` functionality. TaskViewIterator is also a good candidate for users. Perhaps move the stopping capabilities up in the hierarchy?

Reply:
Indeed, included check of max_total_time in for loop as well